src/baldr/shortcut_recovery.h

```diff
@@ -0,0 +1,231 @@
+        // contracted... NOTE: because we change the speed of the edge in graph enhancer we cant use
+        // speed as a reliable determining factor
+        if (begin_node != edge.endnode() && !edge.is_shortcut() &&
+            edge.forwardaccess() == shortcut->forwardaccess() &&
+            edge.reverseaccess() == shortcut->reverseaccess() && edge.sign() == shortcut->sign() &&
+            edge.use() == shortcut->use() && edge.classification() == shortcut->classification() &&
+            edge.roundabout() == shortcut->roundabout() && edge.link() == shortcut->link() &&
+            edge.toll() == shortcut->toll() && edge.destonly() == shortcut->destonly() &&
+            edge.unpaved() == shortcut->unpaved() && edge.surface() == shortcut->surface()/* &&
+          edge.speed() == shortcut->speed()*/) {
+          // we are going to keep this edge
+          edges.emplace_back(tile->header()->graphid());
+          edges.back().set_id(&edge - tile->directededge(0));
+          // and keep expanding from the end of it
+          current_edge = &edge;
+          begin_node = tile->header()->graphid();
+          begin_node.set_id(node_index);
+          accumulated_length += edge.length();
+          break;
+        }
+      }
+
+      // if we didnt add an edge or we went over the length we failed
+      if (current_edge == nullptr || accumulated_length > shortcut->length()) {
+        LOG_TRACE("Unable to recover shortcut for edgeid " + std::to_string(shortcut_id) +
+                  " | accumulated_length: " + std::to_string(accumulated_length) +
+                  " | shortcut_length: " + std::to_string(shortcut->length()));
+        return {shortcut_id};
+      }
+    }
+
+    // we somehow got to the end via a shorter path
+    if (accumulated_length < shortcut->length()) {
+      LOG_TRACE("Unable to recover shortcut for edgeid (accumulated_length < shortcut->length()) " +
+                std::to_string(shortcut_id) +
+                " | accumulated_length: " + std::to_string(accumulated_length) +
+                " | shortcut_length: " + std::to_string(shortcut->length()));
+      return {shortcut_id};
+    }
+
+    // these edges make up this shortcut
+    return edges;
+  }
+
+  // a place to cache the recovered shortcuts
+  std::unordered_map<uint64_t, std::vector<valhalla::baldr::GraphId>> shortcuts;
+  // a place to keep some stats about the recovery
+  size_t unrecovered;
+  size_t superseded;
+
+public:
+  /**
+   * returns a static instance of the cache after prefilling it. if on the first call,
+   * the reader is nullptr then the cache will not be filled and recovery will be on the fly
+   *
+   * @param reader       the reader used to initialize the cache the first time
+   * @return a filled cache mapping shortcuts to superceeded edges
+   */
+  static shortcut_recovery_t& get_instance(valhalla::baldr::GraphReader* reader = nullptr) {
+    static shortcut_recovery_t cache{reader};
```

Comment:
Should `cache` be `thread_local` so as to not break if used from multiple threads?

Reply:
the cache is read-only which is completely fine, and in c++11 and on we are gauranteed that this initialization will happen only once. https://stackoverflow.com/a/19907903/5251867